#!/bin/sh 

Help()
{
    # Display help

    printf "\nCommand: sh curation_2.0_pipe.sh -f <original fasta> -a <agp> <options>\n\n"
    echo "-h Prints help."
    echo "-f Pass original fasta file with combined haplotypes."
    echo "-a Pass the agp generated by PretextView."
    echo "-s Sexual Chromosome marking the heterogametic individuals (e.g. 'Y' for humans, 'W' for birds)"
    # echo "-p Pass the primary assembly you curated (1 for haplotype 1 (default), 2 for haplotype 2)."
    printf "\n"
}

fasta=""
agpfile=""
hap=""

while getopts ":hf:a:s:" option; do
    case $option in 
        h) #display Help
            Help
            exit;;
        f) #Pass original fasta file 
            fasta=$OPTARG;;
        a) #Pass Pretext generated AGP file of curated assembly
            agpfile=$OPTARG;;
        s) #Pass the option of parsing the mashmap output for pairs
            sexchr=$OPTARG;;
        :) # Handle missing arguments for options requiring them
            echo "Error: Option -${OPTARG} requires an argument." >&2
            exit 1
            ;;
        \?) # Handle invalid options
            echo "Error: Invalid option -${OPTARG}." >&2
            exit 1
            ;;
    esac
done

set -e

if [ -d logs ]
then
    count=`ls logs/* | wc -l`
    trimmed=$(echo "$count" | xargs)
    exec 1<> logs/std.${trimmed}.out
else 
    mkdir -p logs 
    exec 1<> logs/std.0.out
    trimmed=0
fi 

pth=$( cd -- "$( dirname -- "${BASH_SOURCE[0]}" )" &> /dev/null && pwd )
printf "Rapid-curation-2.0 scripts located in: $pth\n"

## Programs/tools
# use_gfastats=/vggpfs/fs3/vgl/store/nbrajuka/gfastats/build/bin/gfastats
# use_seqkit=/vggpfs/fs3/vgl/store/nbrajuka/conda/envs/statistics/bin/seqkit
printf "Dependecies:\nBiopython\npandas\ngfastats\nmashmap\nnatsort" 

# could reasonably put everything in a function and just call for both haps. 


printf "\n\nOriginal assembly: ${fasta} \nPretextView generated AGP: ${agpfile}\n\n" ### but checks/breakpoints for if these aren't provided.

printf "Running AGPcorrect on the PretextView generated agp to correct for sequence lengths.\n\n" 
printf "python3 ${pth}/AGPcorrect.py ${fasta} ${agpfile}\n\n"

python3 $pth/AGPcorrect.py ${fasta} ${agpfile} > corrected.agp 

printf "\n\nSplitting the haplotypes from the corrected AGP. Outputs sent to respective directories.\n"
printf "python3 $pth/hap_split.py -1 Hap_1/hap.agp -2 Hap_2/hap.agp -a corrected.agp\n\n"
python3 $pth/hap_split.py  -1 Hap_1/hap.agp -2 Hap_2/hap.agp -a corrected.agp   

printf "\n\nAssigning unlocs before the agp is imposed on the fasta.\n\n"
## If the --agp-to-path in the next block is run first the unlocs will get assimilated into their main assigned scaffolds - they need to be differentiated first.
printf "python3 $pth/unloc.py -a Hap_1/hap.agp -o Hap_1 \n\n" 
python3 $pth/unloc.py -a Hap_1/hap.agp -o Hap_1 
printf "\n\npython3 $pth/unloc.py -a Hap_2/hap.agp -o Hap_2 \n\n"
python3 $pth/unloc.py -a Hap_2/hap.agp -o Hap_2 

printf "\n\nImposing the haplotypic agp on the original fasta to generate a curated fasta.\n"
printf "gfastats $fasta --agp-to-path Hap_1/hap.unlocs.no_hapdups.agp --sort largest -o Hap_1/hap.sorted.fa\n"
gfastats $fasta --agp-to-path Hap_1/hap.unlocs.no_hapdups.agp -o Hap_1/hap.unlocs.no_hapdups.fa  2>> logs/std.agp_to_path_hap1.${trimmed}.out 

printf "\n\ngfastats $fasta --agp-to-path Hap_1/hap.unlocs.no_hapdups.agp --sort largest -o Hap_2/hap.sorted.fa\n\n"
gfastats $fasta --agp-to-path Hap_2/hap.unlocs.no_hapdups.agp -o Hap_2/hap.unlocs.no_hapdups.fa  2>> logs/std.agp_to_path_hap2.${trimmed}.out 

printf "\n\ngfastats Hap_1/hap.unloc.no_hapdups.fa --sort largest -o Hap_1/hap.sorted.fa"
gfastats Hap_1/hap.unlocs.no_hapdups.fa --sort largest -o Hap_1/hap.sorted.fa 

printf "\n\ngfastats Hap_2/hap.unloc.no_hapdups.fa --sort largest -o Hap_2/hap.sorted.fa"
gfastats Hap_2/hap.unlocs.no_hapdups.fa --sort largest -o Hap_2/hap.sorted.fa 

printf "\n\nSubstituting scaffold for chromosome assignments.\n"
printf "python3 $pth/chromosome_assignment.py -a Hap_1/hap.unlocs.no_hapdups.agp -f Hap_1/hap1_sorted.fasta -o Hap_1 \n\n"
python3 $pth/chromosome_assignment.py -a Hap_1/hap.unlocs.no_hapdups.agp -f Hap_1/hap.sorted.fa -o Hap_1 

printf "\n\npython3 $pth/chromosome_assignment.py -a Hap_2/hap.unlocs.no_hapdups.agp -f Hap_2/hap2_sorted.fasta -o Hap_2\n\n"
python3 $pth/chromosome_assignment.py -a Hap_2/hap.unlocs.no_hapdups.agp -f Hap_2/hap.sorted.fa -o Hap_2 



printf "\n\nmashmap -r Hap_1/hap.chr_level.fa  -q Hap_2/hap.chr_level.fa -f one-to-one -t 16 -s 50000 --pi 90 -o mashmap_original_fastas.out \n\n"
mashmap -r Hap_1/hap.chr_level.fa  -q Hap_2/hap.chr_level.fa -f one-to-one -t 16 -s 50000 --pi 90 -o mashmap_original_fastas.out  2>> logs/std.mashmap.${trimmed}.out  

printf "\n\nProcess Mashmap output for reorientation and renaming \n\n"

printf "Give temporary names to hap2 scaffolds for easier renaming : awk '/^>/ {$1 = $1 "_oldname"} {print}' Hap_2/hap.chr_level.fa > Hap_2/tmpnamed.fa \n\n"
awk '/^>/ {$1 = $1 "_oldname"} {print}' Hap_2/hap.chr_level.fa > Hap_2/tmpnamed.fa 

printf "\n\npython3 $pth/filter_mashmap_with_tagged_pairs.py  -1 Hap_1/inter_chr.tsv -2 Hap_2/inter_chr.tsv -r Hap_1 -q Hap_2 -a corrected.agp -m mashmap_original_fastas.out -s ${sexchr}  -o tagged_pairs/ \n \n"
python $pth/filter_mashmap_with_tagged_pairs.py  -1 Hap_1/inter_chr.tsv -2 Hap_2/inter_chr.tsv -r Hap_1 -q Hap_2 -a corrected.agp -m  mashmap_original_fastas.out -s ${sexchr} -o tagged_pairs/ 
    

printf "\n\ngfastats Hap_2/tmpnamed.fa  -k tagged_pairs/reversing_renaming.sak  -o Hap_2/hap2.partially_renamed.fasta \n\n"
gfastats Hap_2/tmpnamed.fa  -k tagged_pairs/reversing_renaming.sak  -o Hap_2/hap2.partially_renamed.fasta 

printf "\n\nremove temporary suffix : sed '/^>/ s/_oldname//' Hap_2/hap2.partially_renamed.fasta >  Hap_2/hap2.reoriented.renamed.fasta \n\n"
sed '/^>/ s/_oldname//' Hap_2/hap2.partially_renamed.fasta >  Hap_2/hap2.reoriented.renamed.fasta 

printf "\n\nSuccessful!\n\n"


exec 1>&-

